crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -71,64 +88,246 @@ impl WasiRuntime {
-            .call(&[])
-            .unwrap();
-
-        info!("module run complete");
-        Ok(())
+            }?;
+
+            info!("module run complete");
+            status_sender
+                .broadcast(ContainerState {
+                    terminated: Some(ContainerStateTerminated {
+                        message: Some("Module run completed".into()),
+                        exit_code: 0,
+                        finished_at: Some(Time(chrono::Utc::now())),
+                        ..Default::default()
+                    }),
+                    ..Default::default()
+                })
+                .expect("status should be able to send");
+            Ok(())
+        });
+
+        Ok(RuntimeHandle::new(
+            tokio::fs::File::from_std(self.output.reopen()?),
+            handle,
+            status_recv,
+        ))
+    }
+}
+
+/// Represents a handle to a running WASI instance. Right now, this is
+/// experimental and just for use with the [crate::WasiProvider]. If we like
+/// this pattern, we will expose it as part of the kubelet crate
+pub struct RuntimeHandle<R: AsyncReadExt + AsyncSeekExt + Unpin> {
+    output: BufReader<R>,
+    handle: JoinHandle<Result<(), failure::Error>>,
+    status_channel: Receiver<ContainerState>,
+}
+
+impl<R: AsyncReadExt + AsyncSeekExt + Unpin> RuntimeHandle<R> {
+    /// Create a new handle with the given reader for log output and a handle to
+    /// the running tokio task. The sender part of the channel should be given
+    /// to the running process and the receiver half passed to this constructor
+    /// to be used for reporting current status
+    pub fn new(
+        output: R,
+        handle: JoinHandle<Result<(), failure::Error>>,
+        status_channel: Receiver<ContainerState>,
+    ) -> Self {
+        RuntimeHandle {
+            output: BufReader::new(output),
+            handle,
+            status_channel,
+        }
+    }
+
+    pub async fn output(&mut self) -> Result<Vec<u8>, failure::Error> {
+        let mut output = Vec::new();
+        self.output.read_to_end(&mut output).await?;
+        // Reset the seek location for the next call to read from the file
+        // NOTE: This is a little janky, but the Tokio BufReader does not
+        // implement the AsyncSeek trait
+        self.output.get_mut().seek(SeekFrom::Start(0)).await?;
+        Ok(output)
+    }
+
+    pub async fn stop(&mut self) -> Result<(), failure::Error> {
+        // TODO: Send an actual stop signal once there is support in wasmtime
+        self.wait().await?;
+        unimplemented!("There is currently no way to stop a running wasmtime instance")
+    }
+
+    pub async fn status(&self) -> Result<ContainerState, failure::Error> {
+        // NOTE: For those who modify this in the future, borrow must be as
+        // short lived as possible. We do not use the recv method because it
+        // uses the value each time and blocks on the next call, whereas we want
```

Comment:
Isn't that it blocks on the `send` side? This comment makes it seem like borrow would block which I don't believe is true.

Reply:
`recv` "blocks" in the await sense. According to the docs, borrowing for a long time _can_ create a block though. I'll reword